dsio/xslx.go

```diff
@@ -0,0 +1,319 @@
+	f           *excelize.File
+	st          *dataset.Structure
+	w           io.Writer
+	types       []string
+}
+
+// NewXLSXWriter creates a Writer from a structure and write destination
+func NewXLSXWriter(st *dataset.Structure, w io.Writer) (*XLSXWriter, error) {
+	// TODO - capture error
+	_, types, _ := terribleHackToGetHeaderRowAndTypes(st)
+
+	wr := &XLSXWriter{
+		st:    st,
+		f:     excelize.NewFile(),
+		types: types,
+		w:     w,
+	}
+
+	if fcg, err := dataset.ParseFormatConfigMap(dataset.XLSXDataFormat, st.FormatConfig); err == nil {
+		if opts, ok := fcg.(*dataset.XLSXOptions); ok {
+			wr.sheetName = opts.SheetName
+		}
+	} else {
+		return nil, err
+	}
+
+	if wr.sheetName == "" {
+		wr.sheetName = "Sheet1"
+	}
+
+	idx := wr.f.NewSheet(wr.sheetName)
+	wr.f.SetActiveSheet(idx)
+
+	return wr, nil
+}
+
+// Structure gives this writer's structure
+func (w *XLSXWriter) Structure() *dataset.Structure {
+	return w.st
+}
+
+// WriteEntry writes one XLSX record to the writer
+func (w *XLSXWriter) WriteEntry(ent Entry) error {
+	if arr, ok := ent.Value.([]interface{}); ok {
+		strs, err := encodeStrings(arr)
+		if err != nil {
+			log.Debug(err.Error())
+			return fmt.Errorf("error encoding entry: %s", err.Error())
+		}
+		for i, str := range strs {
+			w.f.SetCellValue(w.sheetName, w.axis(i), str)
+		}
+		w.rowsWritten++
+		return nil
+	}
+	return fmt.Errorf("expected array value to write xlsx row. got: %v", ent)
+}
+
+func (w *XLSXWriter) axis(colIDx int) string {
+	return ColIndexToLetters(colIDx) + strconv.Itoa(w.rowsWritten+1)
+}
+
+// Close finalizes the writer, indicating no more records
+// will be written
+func (w *XLSXWriter) Close() error {
+	_, err := w.f.WriteTo(w.w)
+	return err
+}
+
+func encodeStrings(vs []interface{}) (strs []string, err error) {
+	strs = make([]string, len(vs))
+	for i, v := range vs {
+		if v == nil {
+			continue
+		}
+		switch x := v.(type) {
+		case int:
+			strs[i] = strconv.Itoa(x)
+		case int64:
+			strs[i] = strconv.Itoa(int(x))
+		case float64:
+			strs[i] = strconv.FormatFloat(x, 'f', -1, 64)
+		case bool:
+			strs[i] = strconv.FormatBool(x)
+		case string:
+			strs[i] = x
+		case []interface{}:
+			data, err := json.Marshal(x)
+			if err != nil {
+				return strs, err
+			}
+			strs[i] = string(data)
+		case map[string]interface{}:
+			data, err := json.Marshal(x)
+			if err != nil {
+				return strs, err
+			}
+			strs[i] = string(data)
+		default:
+			return strs, fmt.Errorf("unrecognized encoding type: %#v", v)
+		}
+	}
+	return
+}
+
+// ColIndexToLetters is used to convert a zero based, numeric column
+// indentifier into a character code.
+func ColIndexToLetters(colRef int) string {
+	parts := intToBase26(colRef)
+	return formatColumnName(smooshBase26Slice(parts))
+}
+
+// Get the largestDenominator that is a multiple of a basedDenominator
+// and fits at least once into a given numerator.
+func getLargestDenominator(numerator, multiple, baseDenominator, power int) (int, int) {
```

Comment:
I realize this function is coming from another source, but I find it pretty confusing, and a bit over complicated. How about:

```
// largestLesserExponent returns the largest exponent of base that does not exceed num
// equivalent to pow(floor(log(num, base)), base)
func largestLesserExponent(num, base int) int {
  prev := 1
  exp := base
  while num >= exp {
    prev = exp
    exp = exp * base
  }
  return prev
}
```

Reply:
this is delightful.